src/append_only_data.rs

```diff
@@ -688,6 +680,167 @@ where
     }
 }
 
+macro_rules! check_perm {
+    ($data: ident, $requester: ident, $request: ident) => {
+        if $data
+            .inner
+            .owners
+            .last()
+            .ok_or_else(|| Error::NoSuchData)?
+            .public_key
+            == $requester
+        {
+            Ok(())
+        } else {
+            $data
+                .inner
+                .permissions
+                .last()
+                .ok_or_else(|| Error::NoSuchData)?
+                .check_permissions_for_key($requester, $request)
+        }
+    };
+}
+
+macro_rules! indices {
+    ($data: ident) => {
+        Ok(Indices::new(
+            $data.entry_index(),
+            $data.owners_index(),
+            $data.permissions_index(),
+        ))
+    };
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub enum AData {
+    PubSeq(PubSeqAppendOnlyData),
+    UnpubSeq(UnpubSeqAppendOnlyData),
+    PubUnseq(PubUnseqAppendOnlyData),
+    UnpubUnseq(UnpubUnseqAppendOnlyData),
+}
+
+impl AData {
```

Comment:
Can this be renamed to `Kind` and exported as `ADataKind`, like how we do with `IDataKind`?

Reply:
Sure @m-cat 🙂 but this [post](https://safenetforum.org/t/rfc-54-published-and-unpublished-datatype/28620/17?u=yogesh) here and it's further discussions are making me think if I should wait.

Reply:
Ah I just saw that too. It's fine to leave this for now until we decide on the naming scheme. For the record I think `ADataKind` is more descriptive than `AData`, but maybe it's more verbose than necessary.